src/common/compat.cc

```diff
@@ -0,0 +1,72 @@
+#include <fcntl.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <sys/param.h>
+#include <sys/mount.h>
+#if defined(__linux__) 
+#include <sys/vfs.h>
+#endif
+
+#include "include/compat.h" 
+
+// On FreeBSD, ZFS fallocate always fails since it is considered impossible to
+// reserve space on a COW filesystem. posix_fallocate() returns EINVAL
+// Linux in this case already emulates the reservation in glibc
+// In which case it is allocated manually, and still that is not a real guarantee
+// that a full buffer is allocated on disk, since it could be compressed.
+// To prevent this the written buffer needs to be loaded with random data.
+
+int manual_fallocate(int fd, off_t offset, off_t len) {
+    // Try to manually allocate the buffer
+    char data[1024*128];
+    int r;
+
+    r = lseek(fd, offset, SEEK_SET);
+    if (r == -1)
+	return -errno;
+    for (off_t off = 0; off < len; off += sizeof(data)) {
+      if (off + sizeof(data) > len)
+        r = write(fd, data, len - off);
+      else
+        r = write(fd, data, sizeof(data));
+      if (r == -1) {
+      	 return -errno;
+      }
+    }
+    return 0;
+}
+
+int on_zfs(int basedir_fd) {
+    struct statfs basefs;
+
+    (void)fstatfs(basedir_fd, &basefs);
+    return (basefs.f_type == FS_ZFS_TYPE);
+}
+
+int ceph_posix_fallocate(int fd, off_t offset, off_t len) {
+
+#ifdef HAVE_POSIX_FALLOCATE
+    if (on_zfs(fd)) {
+        // Preallocate the required space manually
+        return -manual_fallocate(fd, offset, len);
+    } else {
+        return -posix_fallocate(fd, offset, len);
+    }
+#elif defined(__APPLE__)
+    int ret;
+    fstore_t store;
+    store.fst_flags = F_ALLOCATECONTIG;
+    store.fst_posmode = F_PEOFPOSMODE;
+    store.fst_offset = offset;
+    store.fst_length = len;
+
+    ret = fcntl(fd, F_PREALLOCATE, &store);
+    if (ret == -1) {
+      ret = -errno;
+    }
+    return ret;
+#endif
```

Comment:
we might want to keep the fallback code handling the platform where posix_fallocate() is not available? see 75b0f7d1ba02b98112d4281f38e924e4aef30720. also, please note that FileStore does not compile if posix_fallocate() is not available or it targeting platform is not MacOS. while BlueStore has a fallback to emulate fallocate. your PR changes the behavior. please note it down in the commit message.

Reply:
@tchaikov 
Yes I noticed that FileStore did not have a fallback, but just blew up building.
So I considered this a bonus.
But I will put this in the commit message